parser/ast/context.go

```diff
@@ -0,0 +1,43 @@
+package ast
+
+import (
+	"context"
+	"sync"
+)
+
+type modulesKey struct{}
+
+func WithModules(ctx context.Context, mods *ModuleLookup) context.Context {
+	return context.WithValue(ctx, modulesKey{}, mods)
+}
+
+func Modules(ctx context.Context) *ModuleLookup {
+	mods, ok := ctx.Value(modulesKey{}).(*ModuleLookup)
+	if !ok {
+		return NewModules()
+	}
+	return mods
+}
+
+type ModuleLookup struct {
+	mods map[string]*Module
+	mu   sync.Mutex
+}
+
+func NewModules() *ModuleLookup {
+	return &ModuleLookup{
+		mods: make(map[string]*Module),
+	}
+}
+
+func (ml *ModuleLookup) Get(filename string) *Module {
+	ml.mu.Lock()
```

Comment:
Would it make sense to use a `RWLock` and use `RLock()` here? Or does it not really matter?

Alternatively, we could use `sync.Map`, though the docs say:

```
The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.
```

Reply:
I can switch to `sync.RWMutex` to keep things simple.

Reply:
Done.